Documentation/manual/kcc-design/customization-user-story.md

```diff
@@ -0,0 +1,138 @@
+    - The author of the long jump state (the customization user)
+
+Technical Requirements:
+- Ability to pass client-side managed velocity values to the KCCMovementEngine for movement in the XYZ plane
+
+- Collision Phase Callbacks, or the ability to read the state of the KCCMovementEngine after executing the move for the frame
+    - Practical Example:
+        - Imagine we have a charge state that starts at a speed of 0 meters per second and then slowly accelerates in a given direction
+        - At one point, our charging character is moving in the positive z axis and has achieved a speed of 15 meters per second
+        - We then call our move function for the frame, but this time we collide head on to a wall
+        - After executing the move function, I should basically be able to read the following data from the KCCMovementEngine Object
+            - “Your desired velocity for the most recent frame was 15, but your actual velocity was lowered to 0 due to the collision phase”
+            - I can then read that 0 value from the KCC Movement Engine, and re-apply it to my Client Side State
+        - This basically keeps the client side velocity and the KCC Movement Engine state in sync, since with this state definition the user should have to Reaccelerate again once more after colliding, they shouldn’t instantly snap back to 15 meters per second if the wall were suddenly removed
+
+- Collision Phase Callbacks Extended
+    - For example, while in the long jump state, we need to check if we collided with someting head on after executing each move
+        - We would then use this information to enter a Mario style “Bonk” state where the player loses control and our character is repelled from the wall (Same UX as Mario Odyssey where if you long jump into a wall mario hits his head and is repelled)
+
+
+
+
+Taking Action:
+
+I plan to get started making a Mario Odyssey style controller with my current toolset that we can then begin porting to OpenKCC while also using it as a way to define the new API
+
+Upon reflecting on my “Customization User” user stories, I think that the API that solves most of my immediate problems might look something like this, where I can set the velocity of the KCC Movement engine either directly or through a callback that the KCCMovementEngine provides. (Proper architecture withheld for brevity)
+
+```c#
+/*
+* Class: SimpleKCC_ClientMover
+* Simple Direct Usage with One Logical State defined in this classs
+*/
+private void FixedUpdate()
+{
+	// Client Object Updates Its Internal Velocity
+	UpdateVelocityPreMove();
+
+	// Set the Movemnt Engine to Client’s Updated Velocity
+	_movementEngine.SetVelocity(this.Velocity);
+
+	// Movement Engine Performs Move Based on the provided velocity
+	_movementEngine.PerformMove();
+
+	// Read Object State Post-Move to Update Client Velocity In Response to Being Groudned or Not
+	var postMoveGroundedState = _movementEngine.GroundedState;
```

Comment:
We could have these returned as an argument of calling `_movementEngine.PerformMove();` by returning a tuple (GroundedState, Collisions).